agent/consul/gateways/controller_gateways.go

```diff
@@ -2,201 +2,563 @@ package gateways
+	// now check if we need to update the gateway status
+	if toUpdate, shouldUpdate := updater.UpdateEntry(); shouldUpdate {
+		r.logger.Debug("persisting gateway status", "gateway", gateway)
+		if err := r.updater.UpdateWithStatus(toUpdate); err != nil {
+			r.logger.Error("error persisting gateway status", "error", err)
+			return err
+		}
+	}
+
+	// next update route statuses
+	for _, toUpdate := range updatedRoutes {
+		r.logger.Debug("persisting route status", "route", toUpdate)
+		if err := r.updater.UpdateWithStatus(toUpdate); err != nil {
+			r.logger.Error("error persisting route status", "error", err)
+			return err
+		}
+	}
+
+	// now update the bound state if it changed
+	if bound == nil || stateIsDirty(bound.(*structs.BoundAPIGatewayConfigEntry), meta.BoundGateway) {
+		r.logger.Debug("persisting gateway state", "state", meta.BoundGateway)
+		if err := r.updater.Update(meta.BoundGateway); err != nil {
+			r.logger.Error("error persisting state", "error", err)
+			return err
+		}
+	}
+
+	return nil
 }
 
-func (r *apiGatewayReconciler) resourceReferenceToBoundRoute(ref structs.ResourceReference) structs.ControlledConfigEntry {
-	//TODO currently have to retrieve from the store to persist parent field on update call, is there a better way to do this?
-	boundRoute, err := r.store.GetConfigEntry(ref.Kind, ref.Name, &ref.EnterpriseMeta)
+func (r apiGatewayReconciler) cleanupRoute(ctx context.Context, req controller.Request, store *state.Store) error {
+	meta, err := getAllGatewayMeta(store)
 	if err != nil {
-		return nil
+		return err
 	}
 
-	switch ref.Kind {
-	case structs.TCPRoute:
-		return boundRoute.(*structs.TCPRouteConfigEntry)
-	case structs.HTTPRoute:
-		return boundRoute.(*structs.HTTPRouteConfigEntry)
+	r.logger.Debug("cleaning up deleted route object", "request", req)
+	for _, toUpdate := range RemoveRoute(requestToResourceRef(req), meta...) {
+		if err := r.updater.Update(toUpdate.BoundGateway); err != nil {
+			return err
+		}
 	}
-
+	r.controller.RemoveTrigger(req)
 	return nil
 }
 
-// ensureBoundGateway copies all relevant data from a gatewayMeta's APIGateway to BoundAPIGateway
-func (r *apiGatewayReconciler) ensureBoundGateway(gw *gatewayMeta) {
-	if gw.BoundGateway == nil {
-		gw.BoundGateway = &structs.BoundAPIGatewayConfigEntry{
-			Kind:           structs.BoundAPIGateway,
-			Name:           gw.Gateway.Name,
-			EnterpriseMeta: gw.Gateway.EnterpriseMeta,
+// Reconcile reconciles Route config entries.
+func (r apiGatewayReconciler) reconcileRoute(ctx context.Context, req controller.Request, store *state.Store, route structs.BoundRoute) error {
+	now := time.Now()
+
+	meta, err := getAllGatewayMeta(store)
+	if err != nil {
+		return err
+	}
+
+	r.logger.Debug("got route reconcile call", "request", req)
+
+	updater := structs.NewStatusUpdater(route)
+	// we clear out the initial status conditions since we're doing a full update
+	// of this route's status
+	updater.ClearConditions()
+
+	ws := memdb.NewWatchSet()
```

Comment:
This watchset acts as a trigger for us to re-reconcile any time some upstream service has a change in its discovery chain so that we can revalidate the route (i.e. check that it still has the proper protocol alignment).